src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/ProxyService.scala

```diff
@@ -34,41 +37,69 @@ case class AccessTokenExpiredException() extends LeoException(s"Your access toke
+class ProxyService(proxyConfig: ProxyConfig,
+                   gdDAO: DataprocDAO,
+                   dbRef: DbReference,
+                   clusterDnsCache: ActorRef,
+                   authProvider: LeoAuthProvider)(implicit val system: ActorSystem, materializer: ActorMaterializer, executionContext: ExecutionContext) extends LazyLogging {
 
   /* Cache for the bearer token and corresponding google user email */
   private val cachedAuth = CacheBuilder.newBuilder()
     .expireAfterWrite(proxyConfig.cacheExpiryTime, TimeUnit.MINUTES)
     .maximumSize(proxyConfig.cacheMaxSize)
     .build(
-      new CacheLoader[String, Future[(WorkbenchEmail, Instant)]] {
+      new CacheLoader[String, Future[(UserInfo, Instant)]] {
         def load(key: String) = {
-          gdDAO.getEmailAndExpirationFromAccessToken(key)
+          gdDAO.getUserInfoAndExpirationFromAccessToken(key)
         }
       }
     )
 
-  /* Ask the cache for the corresponding google email given a token */
-  def getCachedEmailFromToken(token: String): Future[WorkbenchEmail] = {
-    cachedAuth.get(token).map{ case (email, expireTime) => if (expireTime.isAfter(Instant.now)) email else throw AccessTokenExpiredException() }
+  /* Ask the cache for the corresponding user info given a token */
+  def getCachedUserInfoFromToken(token: String): Future[UserInfo] = {
+    cachedAuth.get(token).map {
+      case (userInfo, expireTime) =>
+        if (expireTime.isAfter(Instant.now))
+          userInfo.copy(tokenExpiresIn = expireTime.toEpochMilli - Instant.now.toEpochMilli)
+        else
+          throw AccessTokenExpiredException() }
   }
 
-
   /**
     * Entry point to this class. Given a google project, cluster name, and HTTP request,
     * looks up the notebook server IP and proxies the HTTP request to the notebook server.
     * Returns NotFound if a notebook server IP could not be found for the project/cluster name.
+    * @param userInfo the current user
     * @param googleProject the Google project
     * @param clusterName the cluster name
     * @param request the HTTP request to proxy
     * @param token the user access token
     * @return HttpResponse future representing the proxied response, or NotFound if a notebook
     *         server IP could not be found.
     */
-  def proxy(googleProject: GoogleProject, clusterName: ClusterName, request: HttpRequest, token: HttpCookiePair): Future[HttpResponse] = {
-    // TODO: in the future we will call sam to do authorization based on the token
-    logger.debug(s"Recevied proxy request with user token ${token.value}")
-    getTargetHost(googleProject, clusterName) flatMap {
+  def proxy(userInfo: UserInfo, googleProject: GoogleProject, clusterName: ClusterName, request: HttpRequest, token: HttpCookiePair): Future[HttpResponse] = {
+    //look up the cluster and check auth to see it
+    val lookupClusterF = dbRef.inTransaction { da => da.clusterQuery.getActiveClusterByName(googleProject, clusterName) }
```

Comment:
I don't like having to do a database query for every call to the Jupyter proxy. Is the only reason to look up the google ID to pass to the auth provider? Can the auth provider accept (googleProject, clusterName) which is also a unique key?

Reply:
Yeah I wasn't keen either. Have pinged this via collabs.